Instantiate a RemoteFrame upon cross-origin iframe navigation when site isolation is enabled

achristensen07 opened the pull request:
#### b6353876dac02d9eca484847c6990996244f103a
<pre>
Instantiate a RemoteFrame upon cross-origin iframe navigation when site isolation is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=249261">https://bugs.webkit.org/show_bug.cgi?id=249261</a>

Reviewed by Tim Horton and Chris Dumez.

This introduces RemoteFrameClient which is kind of like the RemoteFrame equivalent of FrameLoaderClient
for when a RemoteFrame needs to tell the layer above it to do things.  Right now it doesn&apos;t have any actions,
but it will.  It&apos;s also needed for the non-local implementation of WebFrame::fromCoreFrame.

HTMLFrameOwnerElement needs to be able to own a RemoteFrame, so it now owns an AbstractFrame.
Like elsewhere where we do dynamicDowncast&lt;LocalFrame&gt;, there is work yet to do, but that will be done later.

A RemoteFrame needs to own a RemoteDOMWindow like a Frame owns a Document which owns a DOMWindow.
To make this ownership work, I changed the RefPtr&lt;RemoteFrame&gt; to a WeakPtr&lt;RemoteFrame&gt; in RemoteDOMWindow
to prevent reference cycles.

In WebFrame::didCommitLoadInAnotherProcess, I disconnect the Frame and replace it with a RemoteFrame that
is hooked up well enough to pass the existing tests.  More tests need to be written and work needs to be done
there still.

In order to have most of the data available for the tests that use WKWebView._frames to verify things about
site isolation and processes, I had to make frameTreeNodeData take an AbstractFrame so it can return some
meaningful data.  Like almost everything in this PR, there&apos;s still room to fix things there, too!

I wrote an API test that doesn&apos;t really add coverage to this new work that wasn&apos;t already there, but it
sets up for the work I was trying to do when I found I needed to do this first.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::didAttachRenderers):
(WebCore::HTMLFrameElementBase::setFocus):
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::setContentFrame):
(WebCore::HTMLFrameOwnerElement::disconnectContentFrame):
(WebCore::HTMLFrameOwnerElement::contentDocument const):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::HTMLFrameOwnerElement::contentFrame const):
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::evaluateScriptInExtensionTab):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Source/WebCore/page/AbstractFrame.cpp:
(WebCore::parentFrame):
(WebCore::AbstractFrame::AbstractFrame):
(WebCore::AbstractFrame::ownerElement const):
(WebCore::AbstractFrame::disconnectOwnerElement):
* Source/WebCore/page/AbstractFrame.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::contentFrameForNode):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusNavigationScope::scopeOwnedByIFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::advanceFocusDirectionallyInContainer):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::Frame):
(WebCore::Frame::invalidateContentEventRegionsIfNeeded):
(WebCore::Frame::ownerRenderer const):
(WebCore::Frame::contentFrameFromWindowOrFrameElement):
(WebCore::Frame::ownerElement const): Deleted.
(WebCore::Frame::disconnectOwnerElement): Deleted.
(WebCore::Frame::didFinishLoadInAnotherProcess): Deleted.
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/FrameTree.h:
* Source/WebCore/page/ModalContainerObserver.cpp:
(WebCore::ModalContainerObserver::updateModalContainerIfNeeded):
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::appendCustomAttributes):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::RemoteDOMWindow):
(WebCore::RemoteDOMWindow::~RemoteDOMWindow): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::RemoteFrame):
(WebCore::m_client):
(WebCore::RemoteFrame::virtualWindow const):
(WebCore::RemoteFrame::window const):
(WebCore::RemoteFrame::didFinishLoadInAnotherProcess):
(WebCore::RemoteFrame::setWindow): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebCore::RemoteFrameClient::~RemoteFrameClient):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::resolveChildFrameWithNodeHandle):
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
(WebKit::InjectedBundleNodeHandle::htmlFrameElementContentFrame):
(WebKit::InjectedBundleNodeHandle::htmlIFrameElementContentFrame):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebKit::WebRemoteFrameClient::WebRemoteFrameClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebKit::WebRemoteFrameClient::webFrame const):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::fromCoreFrame):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
(WebKit::WebFrame::didFinishLoadInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::frameTreeNodeData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::enableSiteIsolation):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257833@main">https://commits.webkit.org/257833@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fabeacf26b3199ac98593f99d7f60df820b4ad95

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100157 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109490 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169725 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10209 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92585 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107379 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105926 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7718 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90998 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34418 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89624 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22393 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3089 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23908 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3064 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9194 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43394 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5379 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4899 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebKit {
+
+class WebRemoteFrameClient : public WebCore::RemoteFrameClient {
```
final

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -256,14 +262,38 @@ void WebFrame::continueWillSubmitForm(FormSubmitListenerIdentifier listenerID)
+    coreFrame->disconnectOwnerElement();
+    auto client = makeUniqueRef<WebRemoteFrameClient>(*this);
+    auto newFrame = WebCore::RemoteFrame::create(*corePage, m_frameID, ownerElement.get(), WTFMove(client));
+    m_coreFrame = WeakPtr { newFrame.ptr() };
```
Do we really need the explicit `WeakPtr { }`? Also, we should call `.get()` instead of `.ptr()` to avoid an unnecessary null check.

I'd be curious if this would build:
`m_coreFrame = newFrame;`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -36,6 +36,15 @@
+    auto preferences = [configuration preferences];
+    for (_WKExperimentalFeature *feature in [WKPreferences _experimentalFeatures]) {
+        if ([feature.key isEqualToString:@"SiteIsolationEnabled"])
+            [preferences _setEnabled:YES forExperimentalFeature:feature];
```
You can `break;` after this.

achristensen07 committed:
Instantiate a RemoteFrame upon cross-origin iframe navigation when site isolation is enabled
https://bugs.webkit.org/show_bug.cgi?id=249261

Reviewed by Tim Horton and Chris Dumez.

This introduces RemoteFrameClient which is kind of like the RemoteFrame equivalent of FrameLoaderClient
for when a RemoteFrame needs to tell the layer above it to do things.  Right now it doesn't have any actions,
but it will.  It's also needed for the non-local implementation of WebFrame::fromCoreFrame.

HTMLFrameOwnerElement needs to be able to own a RemoteFrame, so it now owns an AbstractFrame.
Like elsewhere where we do dynamicDowncast<LocalFrame>, there is work yet to do, but that will be done later.

A RemoteFrame needs to own a RemoteDOMWindow like a Frame owns a Document which owns a DOMWindow.
To make this ownership work, I changed the RefPtr<RemoteFrame> to a WeakPtr<RemoteFrame> in RemoteDOMWindow
to prevent reference cycles.

In WebFrame::didCommitLoadInAnotherProcess, I disconnect the Frame and replace it with a RemoteFrame that
is hooked up well enough to pass the existing tests.  More tests need to be written and work needs to be done
there still.

In order to have most of the data available for the tests that use WKWebView._frames to verify things about
site isolation and processes, I had to make frameTreeNodeData take an AbstractFrame so it can return some
meaningful data.  Like almost everything in this PR, there's still room to fix things there, too!

I wrote an API test that doesn't really add coverage to this new work that wasn't already there, but it
sets up for the work I was trying to do when I found I needed to do this first.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLFrameElementBase.cpp:
(WebCore::HTMLFrameElementBase::didAttachRenderers):
(WebCore::HTMLFrameElementBase::setFocus):
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::setContentFrame):
(WebCore::HTMLFrameOwnerElement::disconnectContentFrame):
(WebCore::HTMLFrameOwnerElement::contentDocument const):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::HTMLFrameOwnerElement::contentFrame const):
* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::evaluateScriptInExtensionTab):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/SubframeLoader.cpp:
(WebCore::FrameLoader::SubframeLoader::loadOrRedirectSubframe):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::create):
* Source/WebCore/page/AbstractFrame.cpp:
(WebCore::parentFrame):
(WebCore::AbstractFrame::AbstractFrame):
(WebCore::AbstractFrame::ownerElement const):
(WebCore::AbstractFrame::disconnectOwnerElement):
* Source/WebCore/page/AbstractFrame.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::contentFrameForNode):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusNavigationScope::scopeOwnedByIFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::advanceFocusInDocumentOrder):
(WebCore::FocusController::advanceFocusDirectionallyInContainer):
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::Frame):
(WebCore::Frame::invalidateContentEventRegionsIfNeeded):
(WebCore::Frame::ownerRenderer const):
(WebCore::Frame::contentFrameFromWindowOrFrameElement):
(WebCore::Frame::ownerElement const): Deleted.
(WebCore::Frame::disconnectOwnerElement): Deleted.
(WebCore::Frame::didFinishLoadInAnotherProcess): Deleted.
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/FrameTree.h:
* Source/WebCore/page/ModalContainerObserver.cpp:
(WebCore::ModalContainerObserver::updateModalContainerIfNeeded):
* Source/WebCore/page/PageSerializer.cpp:
(WebCore::PageSerializer::SerializerMarkupAccumulator::appendCustomAttributes):
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::RemoteDOMWindow):
(WebCore::RemoteDOMWindow::~RemoteDOMWindow): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::RemoteFrame):
(WebCore::m_client):
(WebCore::RemoteFrame::virtualWindow const):
(WebCore::RemoteFrame::window const):
(WebCore::RemoteFrame::didFinishLoadInAnotherProcess):
(WebCore::RemoteFrame::setWindow): Deleted.
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebCore::RemoteFrameClient::~RemoteFrameClient):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::resolveChildFrameWithNodeHandle):
* Source/WebKit/WebProcess/InjectedBundle/DOM/InjectedBundleNodeHandle.cpp:
(WebKit::InjectedBundleNodeHandle::htmlFrameElementContentFrame):
(WebKit::InjectedBundleNodeHandle::htmlIFrameElementContentFrame):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebKit::WebRemoteFrameClient::WebRemoteFrameClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h: Copied from Source/WebCore/page/RemoteFrame.cpp.
(WebKit::WebRemoteFrameClient::webFrame const):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::fromCoreFrame):
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
(WebKit::WebFrame::didFinishLoadInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::frameTreeNodeData):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::enableSiteIsolation):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257833@main
